src/rules/noMagicNumbersRule.ts

```diff
@@ -63,44 +61,31 @@ export class Rule extends Lint.Rules.AbstractRule {
     public static DEFAULT_ALLOWED = [ -1, 0, 1 ];
 
     public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
-        return this.applyWithWalker(new NoMagicNumbersWalker(sourceFile, this.getOptions()));
+        const options = this.getOptions();
+        const ruleArguments = options.ruleArguments;
+        const allowedNumbers = ruleArguments.length > 0 ? ruleArguments : Rule.DEFAULT_ALLOWED;
+        return this.applyWithWalker(new NoMagicNumbersWalker(sourceFile, options.ruleName, new Set(allowedNumbers.map(String))));
     }
 }
 
-class NoMagicNumbersWalker extends Lint.RuleWalker {
-    // allowed magic numbers
-    private allowed: Set<string>;
-    constructor(sourceFile: ts.SourceFile, options: IOptions) {
-        super(sourceFile, options);
-
-        const configOptions = this.getOptions();
-        const allowedNumbers: number[] = configOptions.length > 0 ? configOptions : Rule.DEFAULT_ALLOWED;
-        this.allowed = new Set(allowedNumbers.map(String));
-    }
-
-    public visitNode(node: ts.Node) {
-        const num = getLiteralNumber(node);
-        if (num !== undefined) {
-            if (!Rule.ALLOWED_NODES.has(node.parent!.kind) && !this.allowed.has(num)) {
-                this.addFailureAtNode(node, Rule.FAILURE_STRING);
+class NoMagicNumbersWalker extends Lint.AbstractWalker<Set<string>> {
+    public walk(sourceFile: ts.SourceFile) {
+        const cb = (node: ts.Node): void => {
+            if (node.kind === ts.SyntaxKind.NumericLiteral) {
+                this.checkNumericLiteral(node, (node as ts.NumericLiteral).text);
+            } else if (node.kind === ts.SyntaxKind.PrefixUnaryExpression &&
+                       (node as ts.PrefixUnaryExpression).operator === ts.SyntaxKind.MinusToken) {
+                this.checkNumericLiteral(node, "-" + ((node as ts.PrefixUnaryExpression).operand as ts.NumericLiteral).text);
+            } else {
+                ts.forEachChild(node, cb);
             }
-        } else {
-            super.visitNode(node);
-        }
+        };
+        return ts.forEachChild(sourceFile, cb);
```

Comment:
That's why the callback function here has the return type annotation `: void`.

The return is actually there on purpose to serve [V8's tail call optimizer](https://github.com/v8/v8/blob/master/src/compiler/tail-call-optimization.cc). AND once ES6 proper tail calls get shipped (currently only available behind `--harmony_tailcalls` flag) we get the ability to handle arbitrarily deep nested trees for free.